lte/gateway/c/oai/tasks/amf/deregistration_request.cpp

```diff
@@ -0,0 +1,300 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef DEREGISTRATION_REQUEST_SEEN
+#define DEREGISTRATION_REQUEST_SEEN
+
+#include <sstream>
+#include <thread>
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include "log.h"
+#include "3gpp_38.401.h"
+#ifdef __cplusplus
+};
+#endif
+#include "common_defs.h"
+#include "amf_app_ue_context_and_proc.h"
+#include "amf_app_defs.h"
+#include "amf_recv.h"
+#include "amf_asDefs.h"
+#include "amf_as.h"
+#include "amf_sap.h"
+#include "amf_app_state_manager.h"
+
+using namespace std;
+namespace magma5g {
+amf_as_data_t amf_data_de_reg_sec;
+
+/*
+ * name : amf_handle_deregistration_ue_origin_req()
+ * Description: Starts processing de-registration request from UE.
+ *        Request comes from AS to AMF as UL NAS message.
+ *        Current scope is 3GPP connection, irrespective of
+ *        switch-off or normal de-registration.
+ *        re-registration required is out of mvc scope now.
+ */
+int amf_handle_deregistration_ue_origin_req(
+    amf_ue_ngap_id_t ue_id, DeRegistrationRequestUEInitMsg* msg, int amf_cause,
+    amf_nas_message_decode_status_t decode_status) {
+  OAILOG_FUNC_IN(LOG_NAS_AMF);
+  OAILOG_DEBUG(
+      LOG_NAS_AMF, "UE originated Deregistration procedures started\n");
+  int rc = RETURNerror;
+  amf_deregistration_request_ies_t params;
+  if (msg->m5gs_de_reg_type.switchoff) {
+    params.de_reg_type = AMF_SWITCHOFF_DEREGISTRATION;
+  } else {
+    params.de_reg_type = AMF_NORMAL_DEREGISTRATION;
+  }
+  /*value of access_type would be 1 or 2 or 3, 24-501 - 9.11.3.20 */
+  switch (msg->m5gs_de_reg_type.access_type) {
+    case 1:
```

Comment:
Use macro `AMF_3GPP_ACCESS` instead of the numeric value 
```suggestion
    case AMF_3GPP_ACCESS:
```

Reply:
Fixed